File: daily/116-Populating-next-right-pointers-in-each-node.py
class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None
        
        def bfs(level_nodes):
            if not level_nodes:
                return
            next_level = []
            for i in range(len(level_nodes)):
                n = level_nodes[i]

                if n.left: 
                    next_level.append(n.left)
                if n.right: 
                    next_level.append(n.right)

                for i in range(len(next_level) - 1):
                    next_level[i].next = next_level[i + 1]

            bfs(next_level)
        bfs([root])
        return root

File: daily/Binary Tree/116-Populating-next-right-pointers-in-each-node.go
package daily

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func connect(root *Node) *Node {
	var bfs func(levelNodes []*Node)
	bfs = func(levelNodes []*Node) {
		if len(levelNodes) == 0 {
			return
		}

		nextLevel := []*Node{}
		for _, n := range levelNodes {
			if n.Left != nil {
				nextLevel = append(nextLevel, n.Left)
			}
			if n.Right != nil {
				nextLevel = append(nextLevel, n.Right)
			}
		}

		for i := 0; i < len(nextLevel)-1; i++ {
			nextLevel[i].Next = nextLevel[i+1]
		}

		bfs(nextLevel)
	}

	if root == nil {
		return nil
	}
	bfs([]*Node{root})
	return root
}
